libbeat/processors/cache/mem_store.go

```diff
@@ -0,0 +1,209 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cache
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+)
+
+// memStore is a memory-backed cache store.
+type memStore struct {
+	mu       sync.Mutex
+	cache    map[string]*CacheEntry
+	expiries expiryHeap
+	ttl      time.Duration // ttl is the time entries are valid for in the cache.
+	refs     int           // refs is the number of processors referring to this store.
+
+	// id is the index into global cache store for the cache.
+	id string
+	// typ is a temporary tag to differentiate memory stores
+	// from the mocked file store.
+	// TODO: Remove when a file-backed store exists.
+	typ string
+
+	// cap is the maximum number of elements the cache
+	// will hold. If not positive, no limit.
+	cap int
+	// effort is the number of entries to examine during
+	// expired element eviction. If not positive, full effort.
+	effort int
+}
+
+// newMemStore returns a new memStore configured to apply the give TTL duration.
+// The memStore is guaranteed not to grow larger than cap elements. id is the
+// look-up into the global cache store the memStore is held in.
+func newMemStore(cfg config, id, typ string) *memStore {
+	return &memStore{
+		id:    id,
+		typ:   typ,
+		cache: make(map[string]*CacheEntry),
+
+		// Mark the ttl as invalid until we have had a put operation
+		// configured.
+		ttl:    -1,
+		cap:    -1,
+		effort: -1,
+	}
+}
+
+func (c *memStore) String() string { return c.typ + ":" + c.id }
+
+// setPutOptions allows concurrency-safe updating of the put options. While the shared
+// backing data store is incomplete, and has no put operation defined, the TTL
+// will be invalid, but will never be accessed since all time operations outside
+// put refer to absolute times. setPutOptions also increases the reference count
+// for the memStore for all operation types.
```

Comment:
> setPutOptions also increases the reference count

That extra behavior was a little surprising. Consider an explicit "retain()" method?

Reply:
Yeah, I was having similar thoughts. This is a consequence of the change to ref counting. I'm reluctant to make this two methods, but I think we get equivalent reduction in confusion by changing the name to `adjust` and removing the comment at the call site.

Reply:
The whole ref count code is now significantly clearer.